torchani/cuaev/aev.h

```diff
@@ -0,0 +1,231 @@
+  Tensor tensor_radialRij;
+  Tensor tensor_angularRij;
+  int total_natom_pairs;
+  int nRadialRij;
+  int nAngularRij;
+  Tensor tensor_centralAtom;
+  Tensor tensor_numPairsPerCenterAtom;
+  Tensor tensor_centerAtomStartIdx;
+  int maxnbrs_per_atom_aligned;
+  int angular_length_aligned;
+  int ncenter_atoms;
+  Tensor coordinates_t;
+  Tensor species_t;
+
+  Result(
+      Tensor aev_t_,
+      Tensor tensor_Rij_,
+      Tensor tensor_radialRij_,
+      Tensor tensor_angularRij_,
+      int64_t total_natom_pairs_,
+      int64_t nRadialRij_,
+      int64_t nAngularRij_,
+      Tensor tensor_centralAtom_,
+      Tensor tensor_numPairsPerCenterAtom_,
+      Tensor tensor_centerAtomStartIdx_,
+      int64_t maxnbrs_per_atom_aligned_,
+      int64_t angular_length_aligned_,
+      int64_t ncenter_atoms_,
+      Tensor coordinates_t_,
+      Tensor species_t_);
+  Result(tensor_list tensors);
+  tensor_list to_list();
+  void release();
+  ~Result() {
+    this->release();
+  }
+};
+
+// cuda kernels
+Result cuaev_forward(const Tensor& coordinates_t, const Tensor& species_t, const AEVScalarParams& aev_params);
+Tensor cuaev_backward(const Tensor& grad_output, const AEVScalarParams& aev_params, const Result& result);
+Tensor cuaev_double_backward(const Tensor& grad_force, const AEVScalarParams& aev_params, const Result& result);
+
+// CuaevComputer
+// Only keep one copy of aev parameters and one copy of result for backward
+struct CuaevComputer : torch::CustomClassHolder {
+  AEVScalarParams aev_params;
+
+  CuaevComputer(
+      double Rcr,
+      double Rca,
+      const Tensor& EtaR_t,
+      const Tensor& ShfR_t,
+      const Tensor& EtaA_t,
+      const Tensor& Zeta_t,
+      const Tensor& ShfA_t,
+      const Tensor& ShfZ_t,
+      int64_t num_species);
+
+  Result forward(const Tensor& coordinates_t, const Tensor& species_t) {
+    return cuaev_forward(coordinates_t, species_t, aev_params);
+  }
+
+  Tensor backward(const Tensor& grad_e_aev, const Result& result) {
+    return cuaev_backward(grad_e_aev, aev_params, result); // force
+  }
+
+  Tensor double_backward(const Tensor& grad_force, const Result& result) {
+    return cuaev_double_backward(grad_force, aev_params, result); // grad_grad_aev
+  }
+};
+
+// Autograd functions
+class CuaevDoubleAutograd : public torch::autograd::Function<CuaevDoubleAutograd> {
+ public:
+  static Tensor forward(
+      AutogradContext* ctx,
+      Tensor grad_e_aev,
+      const torch::intrusive_ptr<CuaevComputer>& cuaev_computer,
+      tensor_list result_tensors);
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+class CuaevAutograd : public torch::autograd::Function<CuaevAutograd> {
+ public:
+  static Tensor forward(
+      AutogradContext* ctx,
+      const Tensor& coordinates_t,
+      const Tensor& species_t,
+      const torch::intrusive_ptr<CuaevComputer>& cuaev_computer);
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
```

Comment:
Should we move them entirely to `cuaev.cpp`? I don't think it is shared between aev.cu and cuaev.cpp, so it does not need to be in header

Reply:
you are right, could this speed up compiling cu file?

Reply:
It will speed up a tiny, probably not observable